Core Metadata v2.1 (#483)
CesiumGS/3d-tiles: specification/Metadata/README.md

- [Overview](#overview)
- [Concepts](#concepts)
- [Schemas](#schemas)
  - [Enums](#enums)
  - [Classes](#classes)
  - [Properties](#properties)
    - [ID](#id)
    - [Name](#name)
    - [Description](#description)
    - [Semantic](#semantic)
    - [Type](#type)
    - [Component Type](#component-type)
    - [Normalized Values](#normalized-values)
    - [Minimum and Maximum Values](#minimum-and-maximum-values)
    - [Required Properties and No Data Values](#required-properties-and-no-data-values)
- [Storage Formats](#storage-formats)
  - [Table Format](#table-format)
    - [Binary Encoding](#binary-encoding)
      - [Numbers](#numbers)
      - [Booleans](#booleans)
      - [Strings](#strings)
      - [Enums](#enums-1)
      - [Arrays](#arrays)
    - [JSON Encoding](#json-encoding)
      - [Numbers](#numbers-1)
      - [Booleans](#booleans-1)
      - [Strings](#strings-1)
      - [Enums](#enums-2)
      - [Arrays](#arrays-1)
  - [Raster Format](#raster-format)
    - [Numbers](#numbers-2)
    - [Enums](#enums-3)
    - [Arrays](#arrays-2)
    - [Implementation Notes](#implementation-notes)
- [Revision History](#revision-history)

## Overview

> The Cesium 3D Metadata Specification defines a standard format for structured metadata in 3D content. Complementary to general-purpose standards like XMP, this specification allows metadata — represented as entities and properties — to be closely associated with parts of 3D content, with data representations appropriate for large, distributed datasets. For the most detailed use cases, properties allow vertex- and texel-level associations; higher-level property associations are also supported.

Reviewer: Yeah could you make a small PR for this in the glTF extension so we don't forget about it?

> Many domains benefit from structured metadata — typical examples include historical details of buildings in a city, names of components in a CAD model, descriptions of regions on textured surfaces, and classification codes for point clouds.

Reviewer: We discussed this earlier this week, we settled on only doing the rename in the glTF extension.

## Concepts

This specification defines metadata schemas and methods for encoding metadata.

**Schemas** contain a set of **classes** and **enums**. Class represents a category of similar entities, defined as a set of **properties**. Each property describes values of a particular type. Enums defines a set of named values representing a single value type, and may be referenced by class properties. Schema definitions do not describe how entities or properties are stored, and may be represented in a file format in various ways. Schemas can be reused across multiple assets or even file formats.

**Entities** are instantiations of class, populated with **property values** conforming to the class definition. Every property value of an entity must be defined by its class, and an entity must not have extraneous property values. Properties of a class may be required, in which case all entities instantiating the class are required to include them.

> >  **Implementation note:** Entities may be defined at various levels of abstraction. Within a large dataset, individual vertices or texels may represent entities with granular metadata properties. The entire dataset may itself be an entity of another class, with its own associated properties.

Reviewer: Yes, I like this alternate wording, please make this change.

#### Name

Names (`name`) provide a human-readable label for a property, and must be unique to a property within a class. Names must be valid UTF-8 strings, and should be written in natural language. Property names do not have inherent meaning; to provide such a meaning, a property must also define a [semantic](#semantics).

> **Example:** A common example of a ID / Name pair would be `localTemperature` and `"Local Temperature"`. Because IDs are restricted to alphanumeric characters and underscores, use of helpful property names is essential for clarity in many languages.

#### Description

Descriptions (`description`) provide a human-readable explanation of a property, its purpose, or its contents. Typically at least a phrase, and possibly several sentences or paragraphs. To provide a machine-readable semantic meaning, a property must also define a [semantic](#semantics).

> #### Semantic

Author: Loosened, making this a recommendation only.

Property IDs, names, and descriptions do not imply semantic meaning within the data format. To provide such a meaning, properties may be assigned a semantic (`semantic`). A semantic is an identifier describing how the property's content should be interpreted, according to an external or application-specific semantic specification.

> > **Example:** Common semantic definitions might include temperature in degrees Celsius, time in milliseconds, or mean squared error (MSE).

Author: I've added a link to `./Semantics/README.md` in the paragraph above – it's part of the same specification, so probably no need for a subset here?

A property's type (`type`) describes the structure of the value given for each entity. Most commonly a single value, a property may also represent a fixed- or variable-length array, or vector and matrix types:

| name   | type                                                    |
|--------|---------------------------------------------------------|
| SINGLE | Single-component value or scalar                        |
| ARRAY  | Fixed- or variable-length array of arbitrary components |
| VEC2   | Fixed-length vector with two (2) numeric components     |
| VEC3   | Fixed-length vector with three (3) numeric components   |
| VEC4   | Fixed-length vector with four (4) numeric components    |
| MAT2   | 2x2 matrix                                              |
| MAT3   | 3x3 matrix                                              |
| MAT4   | 4x4 matrix                                              |

The `ARRAY` type is used to define a fixed- or variable-length array of components. For fixed-length arrays, a component count denotes the number of components in each array, and must be ≥2. Variable-length arrays do not define a component count, and arrays may have any length, including zero.

> The `VECN` and `MATN` types represent specific subsets of the fixed-length `ARRAY` type, where `VECN` is a vector with `N` numeric components and `MATN` is an `N x N` matrix with `N²` numeric components. Where applicable, authoring implementations should choose these more specific types to improve data portability, particularly for mathematical types. Schema representations may choose to make component counts for `VECN` and `MATN` types implicit, rather than storing a `componentCount` descriptor for `VECN` and `MATN` types.

Author: Great point on column-major order, definitely want to include that!

I'll just slim down the "data portability" sentence to the line below; I don't think it's necessary to list benefits here.

> Where applicable, authoring implementations should choose these more specific types.

| name    | componentType                                                             |
|---------|---------------------------------------------------------------------------|
| INT8    | Signed integer in the range `[-128, 127]`                                 |
| UINT8   | Unsigned integer in the range `[0, 255]`                                  |
| INT16   | Signed integer in the range `[-32768, 32767]`                             |
| UINT16  | Unsigned integer in the range `[0, 65535]`                                |
| INT32   | Signed integer in the range `[-2147483648, 2147483647]`                   |
| UINT32  | Unsigned integer in the range `[0, 4294967295]`                           |
| INT64   | Signed integer in the range `[-9223372036854775808, 9223372036854775807]` |
| UINT64  | Unsigned integer in the range `[0, 18446744073709551615]`                 |
| FLOAT32 | A number that can be represented as 32-bit IEEE floating point            |

> | FLOAT64 | A number that can be represented as 64-bit IEEE floating point            |

Reviewer: we probably should be clearer that `NaN` and `Infinity` are excluded

Author: Added the restriction just below the table. 👍

* Unsigned integer values (`UINT8`, `UINT16`, `UINT32`, `UINT64`) must be rescaled to the range `[0.0, 1.0]` (inclusive)
* Signed integer values (`INT8`, `INT16`, `INT32`, `INT64`) must be rescaled to the range `[-1.0, 1.0]` (inclusive)

> **Implementation Note:** Depending on the implementation and the chosen integer type, there may be some loss of precision in the normalized values. For example, if the implementation uses 32-bit floating point variables to represent the normalized value, there are only 23 bits in the mantissa. In this case, if the value to normalize is 32- or 64-bit, a number of lower bits will be truncated when normalized. Therefore, it is recommended that implementations use the highest precision floats available for representing the result.

#### Minimum and Maximum Values

Properties representing numeric values, fixed-length numeric arrays, and vectors may specify a minimum (`minimum`) and maximum (`maximum`). Minimum and maximum values may represent bounds of the valid range for a property, or the exact minimum and maximum values found in the dataset.

> > **Example:** A property storing GPS coordinates might define a range of `[-180, 180]` degrees for longitude values and `[-90, 90]` degrees for latitude values. If the dataset happens to contain only GPS coordinates in a small region, more specific ranges may be given instead.

Reviewer: ```suggestion
> **Example:** A property storing GPS coordinates might define a range of `[-180, 180]` degrees for longitude values and `[-90, 90]` degrees for latitude values. If the dataset contains GPS coordinates only in a small region, more specific ranges may be given instead.
```

Individual components in an array cannot be marked as optional; only the array property itself can be marked as optional.

For array types, `noData` is an array-typed value indicating that the entire array represents a missing value. For example, `[-1, -1, -1]` might be used as a `noData` value for a `VEC3` property. When an array-typed property is required or includes a `noData` value, this has no effect on the interpretation of individual array elements. When variable-length arrays are required, an empty array is still valid.

> **Example:** In the example below, a "tree" class is defined with `noData` indicating a specific enum value to be interpreted as missing data.
>
> | property  | componentType | required | noData    |
> |:----------|:--------------|:---------|:----------|
> | height    | "FLOAT32"     | ✓        |           |

> > | species   | "ENUM"        |          | "Unknown" |

Reviewer: Oh in the paragraph above you mention array types but not enums (which here use the name, rather than the integer value.

Author: Done ✅

> | leafColor | "STRING"      | ✓        |           |

## Storage Formats

A schema provides the pattern for creating entities. This section covers the various formats and encodings for storing entity metadata. Additional formats and encoding may be defined outside of this specification.

* **Table format** - property values are stored in parallel 1D arrays

> * **Raster format** - property values are stored in channels of a 2D grid of pixels

Reviewer: I think we briefly talked about this, could we word this in a way that generalizes to 3D or N-D rasters?

Author: Rephrased: 

* **Raster format** - property values are stored in channels of pixel-based formats (usually images)

... The raster format is for storing fine-grained metadata in pixel-based formats such as images and video.

> - [Overview](#overview)

Reviewer: Please add yourself to the contributors list above between Sam and Bao, thanks!

#### Name

Names (`name`) provide a human-readable label for a property, and must be unique to a property within a class. Names must be valid UTF-8 strings, and should be written in natural language. Property names do not have inherent meaning; to provide such a meaning, a property must also define a [semantic](#semantics).

> **Example:** A typical ID / Name pair, in English, would be `localTemperature` and `"Local Temperature"`. In Japanese, the name might be represented as "きおん". Because IDs are restricted to alphanumeric characters and underscores, use of helpful property names is essential for clarity in many languages.

> **Example:**

#### Description

Descriptions (`description`) provide a human-readable explanation of a property, its purpose, or its contents. Typically at least a phrase, and possibly several sentences or paragraphs. To provide a machine-readable semantic meaning, a property must also define a [semantic](#semantics).

#### Semantic

> Property IDs, names, and descriptions do not impute meaning. To provide such a meaning, properties may be assigned a semantic identifier string (`semantic`), indicating how the property's content should be interpreted. Semantic identifiers may be defined by the [Cesium Metadata Semantic Reference](./Semantics/) or by external semantic references, and may be application-specific. Identifiers should be uppercase, with underscores as word separators.

Reviewer: Comparing the Semantics link to the rest of the document, it looks like we've been inconsistent with the links. I _think_ it's `./Semantics/` like you have here, but good to double check.

### Enums

> An enum consists of a set of named values, represented as `string: integer` pairs. The following enum value types are supported: `INT8`, `UINT8`, `INT16`, `UINT16`, `INT32`, `UINT32`, `INT64`, and `UINT64`. See [Property Types](#type) for definitions of each. Smaller enum types limit the range of possible enum values, and allow more efficient binary encoding. For unsigned value types, enum values most be non-negative. Duplicate names or values within the same enum are not allowed.

Reviewer: I would also change the name of the link to be closer to the section heading for clarity
```suggestion
An enum consists of a set of named values, represented as `string: integer` pairs. The following enum value types are supported: `INT8`, `UINT8`, `INT16`, `UINT16`, `INT32`, `UINT32`, `INT64`, and `UINT64`. See the [Type](#type) section for definitions of each. Smaller enum types limit the range of possible enum values, and allow more efficient binary encoding. For unsigned value types, enum values most be non-negative. Duplicate names or values within the same enum are not allowed.
```

Numeric types are encoded as follows:

| Name    | Description                            |
|---------|----------------------------------------|
| INT8    | 8-bit two's complement signed integer  |
| UINT8   | 8-bit unsigned integer                 |
| INT16   | 16-bit two's complement signed integer |
| UINT16  | 16-bit unsigned integer                |
| INT32   | 32-bit two's complement signed integer |
| UINT32  | 32-bit unsigned integer                |
| INT64   | 64-bit two's complement signed integer |
| UINT64  | 64-bit unsigned integer                |

> | FLOAT32 | 32-bit IEEE floating point             |
> | FLOAT64 | 64-bit IEEE floating point             |

Reviewer: ```suggestion
| FLOAT32 | 32-bit IEEE floating point number      |
| FLOAT64 | 64-bit IEEE floating point number      |
```

<img src="figures/array-of-ints.png"  alt="Variable-length array" width="640px">

**schema → classes → arrayExample**

| property        | type    | componentType |
|-----------------|---------|---------------|
| "arrayProperty" | "ARRAY" | "UINT8"       |

**table**

- count: `5`
- class: `"arrayExample"`
- properties
  - `"arrayProperty"`
    - buffer: `[1, 3, 5, ...]`
    - arrayOffsetBuffer: `[0, 3, 7, ...]`

This example shows how to encode a variable-length array of strings.

![Variable-length array of string](figures/array-of-strings.png)

**schema → classes → weather**

| property          | type    | componentType |
|-------------------|---------|---------------|
| "forecastHistory" | "ARRAY" | "STRING"      |

**table**

- count: `2`
- class `"weather"`
- properties
  - `"forecastHistory"`

>     - buffer: `[R, a, i, n, ...]`

Reviewer: Maybe be more explicit that the buffer stores the individual characters? not sure the best way to indicate this, here's one idea:

```suggestion
    - buffer: `['R', 'a', 'i', 'n', ...]`
```
